### PR TITLE
[Feature] Support unified catalog

### PR DESCRIPTION
**Pull request by letian-jiang**
A unified catalog to handle all types of tables in a single metadata service (e.g. hive-metastore, glue).

```
create external catalog unified 
PROPERTIES ( 
"type" = "unified", 
"unified.metastore.type" = "hive", 
"hive.metastore.uris" = "thrift://{hms_ip}:9083"
);
```

Creating table requires an engine clause. For example, 
```
create table iceberg_table(k1 int, k2 int) engine = iceberg;
```

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Map;
+
+import static com.starrocks.catalog.Table.TableType.DELTALAKE;
+import static com.starrocks.catalog.Table.TableType.HIVE;
+import static com.starrocks.catalog.Table.TableType.HUDI;
+import static com.starrocks.catalog.Table.TableType.ICEBERG;
+
+public class UnifiedConnector implements Connector {
+    public static final String HIVE_METASTORE_URIS = "hive.metastore.uris";
+    public static final String HIVE_METASTORE_TYPE = "hive.metastore.type";
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:41]: 41-44 unused variables

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    private final Map<String, String> properties;
+    private final String catalogName;
+    private final Map<Table.TableType, Connector> connectorMap;
+
+    public UnifiedConnector(ConnectorContext context) {
+        this.properties = context.getProperties();
+        this.catalogName = context.getCatalogName();
+        this.connectorMap = ImmutableMap.of(
+                HIVE, new HiveConnector(context),
+                ICEBERG, new IcebergConnector(context),
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:52]: How to successfully create a unified catalog if you have not set the iceberg parameter.

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 5 / 115 (04.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/HiveTable.java|0|1|00.00%|[171]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedConnector.java|0|14|00.00%|[42, 47, 48, 49, 50, 56, 60, 61, 63, 68, 69, 73, 74, 78]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedMetadata.java|0|94|00.00%|[73, 77, 83, 84, 85, 86, 87, 88, 91, 92, 93, 95, 96, 98, 99, 101, 102, 104, 108, 112, 113, 117, 118, 123, 128, 133, 134, 140, 141, 146, 147, 152, 153, 159, 160, 165, 166, 172, 173, 178, 179, 183, 184, 185, 189, 190, 194, 199, 200, 204, 205, 209, 214, 219, 224, 225, 228, 229, 234, 235, 236, 240, 241, 242, 246, 247, 248, 252, 253, 254, 258, 259, 260, 264, 265, 266, 270, 276, 277, 278, 282, 283, 284, 288, 289, 290, 294, 299, 304, 310, 316, 321, 326, 331]|
|:large_blue_circle:|com/starrocks/connector/ConnectorType.java|4|5|80.00%|[96]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreApiConverter.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 2 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:58]: use `unified.metastore.type`

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 1 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java]: removed

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 14 / 101 (13.86%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/DesensitizedSQLBuilder.java|0|2|00.00%|[472, 475]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedConnector.java|0|18|00.00%|[44, 47, 48, 49, 50, 51, 53, 54, 56, 62, 66, 67, 69, 74, 75, 79, 80, 84]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2899, 2902]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadPendingTask.java|0|1|00.00%|[615]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedMetadata.java|0|61|00.00%|[56, 60, 66, 67, 68, 69, 70, 71, 74, 75, 76, 78, 79, 81, 82, 84, 85, 87, 91, 95, 96, 100, 101, 106, 111, 116, 117, 123, 124, 129, 130, 135, 136, 142, 143, 148, 149, 155, 156, 161, 162, 166, 167, 168, 172, 173, 177, 182, 183, 187, 188, 192, 197, 198, 199, 204, 205, 206, 210, 211, 212]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|7|9|77.78%|[102, 105]|
|:large_blue_circle:|com/starrocks/connector/ConnectorType.java|4|5|80.00%|[96]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreApiConverter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/HiveTableSink.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by stephen-shelby** [comment]
"unified.catalog.type" = "hive" -> "unified.catalog.type" = "hms" ?

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import static com.starrocks.connector.hive.HiveConnector.HIVE_METASTORE_TYPE;
+import static com.starrocks.connector.iceberg.IcebergConnector.ICEBERG_CATALOG_TYPE;
+
+public class UnifiedConnector implements Connector {
+    public static final String UNIFIED_METASTORE_TYPE = "unified.metastore.type";
+    public static final List<String> SUPPORTED_METASTORE_TYPE = ImmutableList.of("hive", "glue");
+    private final Map<Table.TableType, Connector> connectorMap;
+
+    public UnifiedConnector(ConnectorContext context) {
+        String metastoreType = context.getProperties().get(UNIFIED_METASTORE_TYPE);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:48]: check null

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        if (!SUPPORTED_METASTORE_TYPE.contains(metastoreType)) {
+            throw new SemanticException("Unified catalog only supports hive and glue as metastore.");
+        }
+
+        ImmutableMap.Builder<String, String> derivedProperties = ImmutableMap.builder();
+        derivedProperties.putAll(context.getProperties());
+        derivedProperties.put(HIVE_METASTORE_TYPE, metastoreType);
+        derivedProperties.put(ICEBERG_CATALOG_TYPE, metastoreType);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:56]: Does paimon need to be compatible?

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    public void bindConfig(ConnectorConfig config) {
+        connectorMap.forEach((k, v) -> v.bindConfig(config));
+    }
+
+    @Override
+    public CloudConfiguration getCloudConfiguration() {
+        return connectorMap.get(HIVE).getCloudConfiguration();
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:89]: why only use HIVE?

**Comment by letian-jiang** [comment]
> "unified.catalog.type" = "hive" -> "unified.catalog.type" = "hms" ?

Sorry for the mistype. Actually `unified.metastore.type` is used to specify external metadata service.

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        requireNonNull(metadataMap, "metadataMap is null");
+        checkArgument(metadataMap.containsKey(HIVE), "metadataMap does not have hive metadata");
+        this.metadataMap = metadataMap;
+        this.hiveMetadata = (HiveMetadata) metadataMap.get(HIVE);
+    }
+
+    private Table.TableType getTableType(String dbName, String tblName) {
+        Table table = hiveMetadata.getTable(dbName, tblName);
+        if (table == null) {
+            return HIVE; // use hive metadata by default
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:76]: what's the scenes that table from hms is null?

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        ImmutableMap.Builder<String, String> derivedProperties = ImmutableMap.builder();
+        derivedProperties.putAll(context.getProperties());
+        derivedProperties.put(HIVE_METASTORE_TYPE, metastoreType);
+        derivedProperties.put(ICEBERG_CATALOG_TYPE, metastoreType);
+
+        ConnectorContext derivedContext = new ConnectorContext(context.getCatalogName(), context.getType(),
+                derivedProperties.build());
+
+        connectorMap = ImmutableMap.of(
+                HIVE, new HiveConnector(derivedContext),
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:62]: you need consider CatalogConnector

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    @Override
+    public boolean dbExists(String dbName) {
+        return hiveMetadata.dbExists(dbName);
+    }
+
+    @Override
+    public void createDb(String dbName, Map<String, String> properties) throws DdlException, AlreadyExistsException {
+        hiveMetadata.createDb(dbName, properties);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:182]: iceberg catalog can't create db by hms

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java (hunk 6 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:76]: Refer to HiveMetadata#getTable

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 5 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:89]: Cloud configurations are the same for any sub-connector

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 7 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:62]: No need. They are different levels.

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 4 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:56]: Nope.

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java (hunk 8 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:182]: I think creating db is irrelevant to specifical catalog type

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java (hunk 3 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedConnector.java:48]: null check has been performed at the next line

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+    public UnifiedMetadata(Map<Table.TableType, ConnectorMetadata> metadataMap) {
+        requireNonNull(metadataMap, "metadataMap is null");
+        checkArgument(metadataMap.containsKey(HIVE), "metadataMap does not have hive metadata");
+        this.metadataMap = metadataMap;
+        this.hiveMetadata = (HiveMetadata) metadataMap.get(HIVE);
+    }
+
+    private Table.TableType getTableType(String dbName, String tblName) {
+        Table table = hiveMetadata.getTable(dbName, tblName);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:74]: if tblname is a iceberg/delta lake table, hiveMetadata.getTable could get iceberg/delta lake table?

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    public static final String ICEBERG_TABLE_TYPE_VALUE = "iceberg";
+    public static final String SPARK_TABLE_PROVIDER_KEY = "spark.sql.sources.provider";
+    public static final String DELTA_LAKE_PROVIDER = "delta";
+
+    static boolean isIcebergTable(Map<String, String> properties) {
+        return ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(properties.get(ICEBERG_TABLE_TYPE_NAME));
+    }
+
+    static boolean isDeltaLakeTable(Map<String, String> properties) {
+        return DELTA_LAKE_PROVIDER.equalsIgnoreCase(properties.get(SPARK_TABLE_PROVIDER_KEY));
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:60]: only support spark write table？ how about trino，presto?

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java [hunk 11]
```diff
@@ -93,19 +94,26 @@ private static String analyzeEngineName(String engineName, String catalogName) {
...
             if (Strings.isNullOrEmpty(engineName)) {
-                if (!catalogType.equalsIgnoreCase(ICEBERG) && !catalogType.equalsIgnoreCase(HIVE)) {
-                    throw new SemanticException("Currently doesn't support creating tables of type " + catalogType);
-                }
-                return catalogType;
-            } else if (!engineName.equalsIgnoreCase(catalogType)) {
-                throw new SemanticException("Can't create %s table in the %s catalog", engineName, catalogType);
-            } else {
-                return engineName;
+                throw new SemanticException("Create table in unified catalog requires engine clause " +
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:102]: plz give a example how to create a table in unified catalog

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java (hunk 9 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:74]: Yes. Use table properties to determine table type.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java (hunk 11 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/CreateTableAnalyzer.java:102]: add a sql in description

File: fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java (hunk 10 above)
Comment (letian-jiang) [reply fe/fe-core/src/main/java/com/starrocks/connector/unified/UnifiedMetadata.java:60]: This refers to the implementation of trino

```java
public static boolean isDeltaLakeTable(Map<String, String> tableParameters)
{
    return DELTA_LAKE_PROVIDER.equalsIgnoreCase(tableParameters.get(SPARK_TABLE_PROVIDER_KEY));
}
```

I think these engines all conform to this setting.

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 11 / 17 (64.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/DeltaLakeTable.java|0|2|00.00%|[49, 50]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadPendingTask.java|0|1|00.00%|[615]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|[2902]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|8|10|80.00%|[115, 116]|
|:large_blue_circle:|com/starrocks/connector/ConnectorType.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 82 / 104 (78.85%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/dump/DesensitizedSQLBuilder.java|0|2|00.00%|[472, 475]|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|[2906, 2909]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadPendingTask.java|0|1|00.00%|[615]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|7|9|77.78%|[102, 105]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedMetadata.java|48|60|80.00%|[75, 148, 149, 154, 155, 185, 190, 191, 192, 197, 198, 199]|
|:large_blue_circle:|com/starrocks/connector/unified/UnifiedConnector.java|17|20|85.00%|[49, 83, 84]|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreApiConverter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/DeltaLakeTable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/ConnectorType.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/HiveTableSink.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by stephen-shelby — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**